solidity/contracts/bridge/library/Fraud.sol

```diff
@@ -0,0 +1,393 @@
+        mapping(uint256 => bool) storage spentMainUTXOs
+    ) external {
+        bytes32 sighash = preimage.hash256();
+
+        uint256 challengeKey = uint256(
+            keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+
+        require(challenge.reportedAt > 0, "Fraud challenge does not exist");
+        require(!challenge.closed, "Fraud challenge closed");
+
+        // Ensure SIGHASH_ALL type was used during signing, which is represented
+        // by type value `1`.
+        require(extractSighashType(preimage) == 1, "Wrong sighash type");
+
+        if (witness) {
+            verifyWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        } else {
+            verifyNonWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        }
+        // If we passed the preimage verification, the wallet has successfully
+        // defended the fraud challenge.
+
+        // Send the ether deposited by the challenger to the treasury
+        /* solhint-disable avoid-low-level-calls */
+        // slither-disable-next-line low-level-calls
+        (bool success, ) = treasury.call{value: challenge.ethDepositAmount}("");
+        require(success, "Failed to send Ether");
+        /* solhint-enable avoid-low-level-calls */
+
+        // TODO: Reward the caller for defeating the challenge successfully.
+        challenge.closed = true;
+
+        // TODO: Consider emitting the event with walletPublicKey in the
+        //       compressed format as it's how we identify wallets in the Bridge.
+        emit FraudChallengeDefeated(walletPublicKey, sighash, v, r, s);
+    }
+
+    // TODO: description
+    function notifyChallengeTimeout(
+        Data storage self,
+        bytes memory walletPublicKey,
+        bytes32 sighash,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        uint256 challengeKey = uint256(
+            keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+        require(challenge.reportedAt > 0, "Fraud challenge does not exist");
+        require(!challenge.closed, "Fraud challenge is closed");
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp - challenge.reportedAt >=
+                self.fraudChallengeDefendTimeout,
+            "Fraud challenge defend timeout has not elapsed"
+        );
+
+        // TODO: Slash the wallet
+        // TODO: Reward the challenger
+
+        challenge.closed = true;
+
+        // Return the ether deposited by the challenger
+        /* solhint-disable avoid-low-level-calls */
+        // slither-disable-next-line low-level-calls
+        (bool success, ) = challenge.challenger.call{
+            value: challenge.ethDepositAmount
+        }("");
+        require(success, "Failed to send Ether");
```

Comment:
I think we should also consider reentrancy here since any address can be the challenger. I don't see any problems atm as all functions follow the check-effect-interactions pattern but it may be worth considering introducing the pull payment mechanism or gas limit for the transfer call just in case, to avoid a possible attack vector in the future.

Reply:
Good catch, done in 1338b7ec902bee21d2a8a13010adf252a6f8e5f1.

Reply:
> I think we should also consider reentrancy here since any address can be the challenger. I don't see any problems atm as all functions follow the check-effect-interactions pattern but it may be worth considering introducing the pull payment mechanism or gas limit for the transfer call just in case, to avoid a possible attack vector in the future.

@lukasz-zimnoch Yes, it's worth adding some gas limit during sending of ether as `call` forwards all the available gas.
Any idea of what the limit value could be? I added `5000` for now. Other functions (`send` and `transfer`) have a limit of `2300`. Maybe we could add a little bit more than that in case there is some fork in ethereum and more gas is required in the future to make ether transfer.
Not sure about the pull payment. 
@pdyraga Any opinion?